From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kihz17 <mitchdawdy@gmail.com>
Date: Mon, 10 Mar 2025 19:46:29 -0700
Subject: [PATCH] fixup! paper File Patches


diff --git a/net/minecraft/network/syncher/SynchedEntityData.java b/net/minecraft/network/syncher/SynchedEntityData.java
index d15d6e114819eeb78cc31f7d4214e8635ed4ecb1..c044a10a8e433f48fd80fd72a34903950ee9b434 100644
--- a/net/minecraft/network/syncher/SynchedEntityData.java
+++ b/net/minecraft/network/syncher/SynchedEntityData.java
@@ -166,6 +166,18 @@ public class SynchedEntityData {
             }
         }
 
+        public <T> SynchedEntityData.Builder forceDefine(EntityDataAccessor<T> accessor, T value) {
+            int id = accessor.id();
+            if (id > this.itemsById.length) {
+                throw new IllegalArgumentException("Data value id is too big with " + id + "! (Max is " + this.itemsById.length + ")");
+            } else if (EntityDataSerializers.getSerializedId(accessor.serializer()) < 0) {
+                throw new IllegalArgumentException("Unregistered serializer " + accessor.serializer() + " for " + id + "!");
+            } else {
+                this.itemsById[accessor.id()] = new SynchedEntityData.DataItem<>(accessor, value);
+                return this;
+            }
+        }
+
         public SynchedEntityData build() {
             for (int i = 0; i < this.itemsById.length; i++) {
                 if (this.itemsById[i] == null) {
